### PR TITLE
Java http dest

### LINKED ISSUE #539: HTTP destination driver in Java
porting HTTP dest driver from Python to Java

### PR DESCRIPTION
**Pull request by aneutrals**
HTTP destination implemented that can use the HTTP `PUT` method to send data to an HTTP server.

fixes #539

### CODE REVIEW

**Comment by lbudai** [comment]
hi,

maybe you forgot to rebase to the actual master because your PR containing some extra python-related patches and not only Java HTTP dest.

**Comment by lbudai** [comment]
dependes on #548

File: modules/java/HTTPDestination.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/**
+ * Created on 6/29/15.
+ */
```
Comment (gregory094) [inline modules/java/HTTPDestination.java]: Please remove this comment above.
Git history will show when this file was added first.

File: modules/java/HTTPDestination.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    public boolean open() {
+        return true;
+    }
+
+    public boolean isOpened() {
+        return true;
+    }
+
+    public void close() {
+
```
Comment (gregory094) [inline modules/java/HTTPDestination.java]: Please remove this empty line in order to ensure the consistency in the code. (like in `deinit()`)

File: modules/java/HTTPDestination.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            connection.setDoOutput(true);
+            OutputStreamWriter osw = new OutputStreamWriter(connection.getOutputStream());
+            osw.write(message);
+            osw.flush();
+            osw.close();
+            responseCode=connection.getResponseCode();
+        }
+        catch (Exception e)
+        {
+
```
Comment (gregory094) [inline modules/java/HTTPDestination.java]: Please remove this empty line in order to ensure the consistency in the code.

File: modules/java/HTTPDestination.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        return true;
+    }
+
+    public void close() {
+
+    }
+
+    public boolean send(String message) {
+        int responseCode=0;
+        try {
```
Comment (gregory094) [inline modules/java/HTTPDestination.java]: Please follow the same bracketing rule in the code. (try-catch blocks)

File: modules/java/HTTPDestination.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            HttpURLConnection connection = (HttpURLConnection) this.url.openConnection();
+            connection.setRequestMethod("PUT");
+            connection.setRequestProperty( "content-length", Integer.toString(message.length()) );
+            connection.setDoOutput(true);
+            OutputStreamWriter osw = new OutputStreamWriter(connection.getOutputStream());
+            osw.write(message);
+            osw.flush();
+            osw.close();
```
Comment (gregory094) [inline modules/java/HTTPDestination.java]: We should move `osw.close()` to the end of the function to ensure that the connection will be closed.
note: create a `finally` block at the end of the `try-catch`

File: modules/java/HTTPDestination.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            responseCode=connection.getResponseCode();
```
Comment (gregory094) [inline modules/java/HTTPDestination.java]: This should be moved to the `catch` block for response code is only needed there.

**Comment by lbudai** [comment]
I think this PR should be rebased to current master and should use the common java library.

**Comment by aneutrals** [comment]
I rebased this to the curernt master and it now uses the common library.

File: modules/java-modules/Makefile.am [hunk 7]
```diff
@@ -75,5 +76,7 @@ EXTRA_DIST += \
         modules/java-modules/common/src/test/java/org/syslog_ng/logging/test/MockLogDestination.java \
         modules/java-modules/hdfs/src/main/java/org/syslog_ng/hdfs/HdfsDestination.java \
         modules/java-modules/hdfs/src/main/java/org/syslog_ng/hdfs/HdfsOptions.java \
+        modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestination.java \
+        modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestinationOptions.java \
         modules/java-modules/dummy/src/main/java/org/syslog_ng/DummyTextDestination.java
```
Comment (lbudai) [inline modules/java-modules/Makefile.am]: Could you fix those indentation differencies?
I know it is a small issue but hurts my eyes :-)

File: modules/java-modules/http/build.gradle [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+project.buildDir = syslogBuildDir+'/http/gradle'
+
+
+repositories {
+    flatDir {
+       dirs syslogDepsDir
+       dirs syslogBuildDir + '/common/http/libs'
+       dirs '/usr/lib/syslog-ng-java-module-dependency-jars/jars'
+    }
```
Comment (lbudai) [inline modules/java-modules/http/build.gradle]: Does HTTP dst needs those dependencies?

File: modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestination.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+
+public class HTTPDestination extends TextLogDestination {
+
+
+    private URL url;
```
Comment (lbudai) [inline modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestination.java]: There are some extra new lines here.

File: modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestination.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    Logger logger;
+    private HTTPDestinationOptions options;
+
+    public HTTPDestination(long arg0) {
+        super(arg0);
```
Comment (lbudai) [inline modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestination.java]: A better name required for `arg0` :-)

File: modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestination.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        int responseCode = 0;
+        try {
+            HttpURLConnection connection = (HttpURLConnection) this.url.openConnection();
+            connection.setRequestMethod("PUT");
+            connection.setRequestProperty("content-length", Integer.toString(message.length()));
+            connection.setDoOutput(true);
+            try (OutputStreamWriter osw = new OutputStreamWriter(connection.getOutputStream())) {
+                osw.write(message);
+                osw.close();
+            } catch (Exception e) {
```
Comment (lbudai) [inline modules/java-modules/http/src/main/java/org/syslog_ng/http/HTTPDestination.java]: Are you sure you want to catch `Exception` ?

**Pull request merged**